File: macos/Sources/WiwyTransfer/QuickShare/InboundNearbyConnection.swift
// Portado/adaptado de NearDrop (The Unlicense): https://github.com/grishka/NearDrop
// Receptor: maneja el handshake UKEY2 (rol servidor), el consentimiento y la
// recepción de archivos, guardándolos en Descargas/WiwyTransfer.

import Foundation
import Network
import CryptoKit
import CommonCrypto

import SwiftECC
import BigInt

protocol InboundNearbyConnectionDelegate: AnyObject {
    func obtainUserConsent(for transfer: TransferMetadata, from device: RemoteDeviceInfo, connection: InboundNearbyConnection)
    func connectionProgress(_ connection: InboundNearbyConnection, received: Int64, total: Int64, currentFile: String)
    func connectionFinished(_ connection: InboundNearbyConnection, savedPaths: [String], from device: RemoteDeviceInfo)
    func connectionWasTerminated(connection: InboundNearbyConnection, error: Error?)
}

final class InboundNearbyConnection: NearbyConnection {

    private var currentState: State = .initial
    weak var delegate: InboundNearbyConnectionDelegate?
    private var cipherCommitment: Data?

    private let saveDirectory: URL
    private var savedPaths: [String] = []
    private var totalBytes: Int64 = 0
    private var receivedBytes: Int64 = 0
    private var currentFileName: String = ""

    enum State {
        case initial, receivedConnectionRequest, sentUkeyServerInit, receivedUkeyClientFinish,
             sentConnectionResponse, sentPairedKeyResult, receivedPairedKeyResult,
             waitingForUserConsent, receivingFiles, disconnected
    }

    init(connection: NWConnection, id: String, saveDirectory: URL) {
        self.saveDirectory = saveDirectory
        super.init(connection: connection, id: id)
    }

    override func handleConnectionClosure() {
        super.handleConnectionClosure()
        let wasReceiving = currentState == .receivingFiles
        currentState = .disconnected
        try? deletePartiallyReceivedFiles()
        DispatchQueue.main.async {
            self.delegate?.connectionWasTerminated(connection: self, error: wasReceiving ? self.lastError : self.lastError)
        }
    }

    override internal func processReceivedFrame(frameData: Data) {
        do {
            switch currentState {
            case .initial:
                let frame = try Location_Nearby_Connections_OfflineFrame(serializedData: frameData)
                try processConnectionRequestFrame(frame)
            case .receivedConnectionRequest:
                let msg = try Securegcm_Ukey2Message(serializedData: frameData)
                ukeyClientInitMsgData = frameData
                try processUkey2ClientInit(msg)
            case .sentUkeyServerInit:
                let msg = try Securegcm_Ukey2Message(serializedData: frameData)
                try processUkey2ClientFinish(msg, raw: frameData)
            case .receivedUkeyClientFinish:
                let frame = try Location_Nearby_Connections_OfflineFrame(serializedData: frameData)
                try processConnectionResponseFrame(frame)
            default:
                let smsg = try Securemessage_SecureMessage(serializedData: frameData)
                try decryptAndProcessReceivedSecureMessage(smsg)
            }
        } catch {
            lastError = error
            print("QS: error de deserialización: \(error) en estado \(currentState)")
            protocolError()
        }
    }

    override internal func processTransferSetupFrame(_ frame: Sharing_Nearby_Frame) throws {
        if frame.hasV1 && frame.v1.hasType, case .cancel = frame.v1.type {
            print("QS: transferencia cancelada")
            try sendDisconnectionAndDisconnect()
            return
        }
        switch currentState {
        case .sentConnectionResponse:
            try processPairedKeyEncryptionFrame(frame)
        case .sentPairedKeyResult:
            try processPairedKeyResultFrame(frame)
        case .receivedPairedKeyResult:
            try processIntroductionFrame(frame)
        default:
            print("QS: estado inesperado en processTransferSetupFrame: \(currentState)")
        }
    }

    override func isServer() -> Bool { true }

    override func processFileChunk(frame: Location_Nearby_Connections_PayloadTransferFrame) throws {
        let id = frame.payloadHeader.id
        guard let fileInfo = transferredFiles[id] else { throw NearbyError.protocolError("ID de payload \(id) desconocido") }
        let currentOffset = fileInfo.bytesTransferred
        guard frame.payloadChunk.offset == currentOffset else { throw NearbyError.protocolError("Offset inválido \(frame.payloadChunk.offset), esperado \(currentOffset)") }
        guard currentOffset + Int64(frame.payloadChunk.body.count) <= fileInfo.meta.size else { throw NearbyError.protocolError("El tamaño del archivo excede lo anunciado") }
        if frame.payloadChunk.body.count > 0 {
            fileInfo.fileHandle?.write(frame.payloadChunk.body)
            transferredFiles[id]!.bytesTransferred += Int64(frame.payloadChunk.body.count)
            receivedBytes += Int64(frame.payloadChunk.body.count)
            reportProgress(fileName: fileInfo.meta.name)
        } else if (frame.payloadChunk.flags & 1) == 1 {
            try fileInfo.fileHandle?.close()
            transferredFiles[id]!.fileHandle = nil
            savedPaths.append(fileInfo.destinationURL.path)
            transferredFiles.removeValue(forKey: id)
            if transferredFiles.isEmpty {
                finishTransfer()
                try sendDisconnectionAndDisconnect()
            }
        }
    }

    override func processBytesPayload(payload: Data, id: Int64) throws -> Bool {
        if let fileInfo = transferredFiles[id] {
            fileInfo.fileHandle?.write(payload)
            transferredFiles[id]!.bytesTransferred += Int64(payload.count)
            receivedBytes += Int64(payload.count)
            reportProgress(fileName: fileInfo.meta.name)
            try fileInfo.fileHandle?.close()
            transferredFiles[id]!.fileHandle = nil
            savedPaths.append(fileInfo.destinationURL.path)
            transferredFiles.removeValue(forKey: id)
            if transferredFiles.isEmpty {
                finishTransfer()
                try sendDisconnectionAndDisconnect()
            }
            return true
        }
        return false
    }

    private func reportProgress(fileName: String) {
        let total = totalBytes
        let received = receivedBytes
        DispatchQueue.main.async {
            self.delegate?.connectionProgress(self, received: received, total: total, currentFile: fileName)
        }
    }

    private func finishTransfer() {
        let paths = savedPaths
        let device = remoteDeviceInfo
        DispatchQueue.main.async {
            if let device = device {
                self.delegate?.connectionFinished(self, savedPaths: paths, from: device)
            }
        }
    }

    private func processConnectionRequestFrame(_ frame: Location_Nearby_Connections_OfflineFrame) throws {
        guard frame.hasV1 && frame.v1.hasConnectionRequest && frame.v1.connectionRequest.hasEndpointInfo else { throw NearbyError.requiredFieldMissing("connectionRequest.endpointInfo") }
        guard case .connectionRequest = frame.v1.type else { throw NearbyError.protocolError("Tipo de frame inesperado \(frame.v1.type)") }
        let endpointInfo = frame.v1.connectionRequest.endpointInfo
        guard endpointInfo.count > 17 else { throw NearbyError.protocolError("Endpoint info demasiado corto") }
        let b = [UInt8](endpointInfo)
        let deviceNameLength = Int(b[17])
        guard b.count >= deviceNameLength + 18 else { throw NearbyError.protocolError("Endpoint info no contiene el nombre") }
        guard let deviceName = String(bytes: b[18..<(18 + deviceNameLength)], encoding: .utf8) else { throw NearbyError.protocolError("Nombre de dispositivo no es UTF-8 válido") }
        let rawDeviceType = Int(b[0] & 7) >> 1
        remoteDeviceInfo = RemoteDeviceInfo(name: deviceName, type: RemoteDeviceInfo.DeviceType.fromRawValue(value: rawDeviceType))
        currentState = .receivedConnectionRequest
    }

    private func processUkey2ClientInit(_ msg: Securegcm_Ukey2Message) throws {
        guard msg.hasMessageType, msg.hasMessageData else { throw NearbyError.requiredFieldMissing("clientInit ukey2message.type|data") }
        guard case .clientInit = msg.messageType else { sendUkey2Alert(type: .badMessageType); throw NearbyError.ukey2 }
        let clientInit: Securegcm_Ukey2ClientInit
        do {
            clientInit = try Securegcm_Ukey2ClientInit(serializedData: msg.messageData)
        } catch {
            sendUkey2Alert(type: .badMessageData); throw NearbyError.ukey2
        }
        guard clientInit.version == 1 else { sendUkey2Alert(type: .badVersion); throw NearbyError.ukey2 }
        guard clientInit.random.count == 32 else { sendUkey2Alert(type: .badRandom); throw NearbyError.ukey2 }
        var found = false
        for commitment in clientInit.cipherCommitments {
            if case .p256Sha512 = commitment.handshakeCipher {
                found = true
                cipherCommitment = commitment.commitment
                break
            }
        }
        guard found else { sendUkey2Alert(type: .badHandshakeCipher); throw NearbyError.ukey2 }
        guard clientInit.nextProtocol == "AES_256_CBC-HMAC_SHA256" else { sendUkey2Alert(type: .badNextProtocol); throw NearbyError.ukey2 }

        let domain = Domain.instance(curve: .EC256r1)
        let (pubKey, privKey) = domain.makeKeyPair()
        publicKey = pubKey
        privateKey = privKey

        var serverInit = Securegcm_Ukey2ServerInit()
        serverInit.version = 1
        serverInit.random = Data.randomData(length: 32)
        serverInit.handshakeCipher = .p256Sha512

        var pkey = Securemessage_GenericPublicKey()
        pkey.type = .ecP256
        pkey.ecP256PublicKey = Securemessage_EcP256PublicKey()
        pkey.ecP256PublicKey.x = Data(pubKey.w.x.asSignedBytes())
        pkey.ecP256PublicKey.y = Data(pubKey.w.y.asSignedBytes())
        serverInit.publicKey = try pkey.serializedData()

        var serverInitMsg = Securegcm_Ukey2Message()
        serverInitMsg.messageType = .serverInit
        serverInitMsg.messageData = try serverInit.serializedData()
        let serverInitData = try serverInitMsg.serializedData()
        ukeyServerInitMsgData = serverInitData
        sendFrameAsync(serverInitData)
        currentState = .sentUkeyServerInit
    }

    private func processUkey2ClientFinish(_ msg: Securegcm_Ukey2Message, raw: Data) throws {
        guard msg.hasMessageType, msg.hasMessageData else { throw NearbyError.requiredFieldMissing("clientFinish ukey2message.type|data") }
        guard case .clientFinish = msg.messageType else { throw NearbyError.ukey2 }

        var sha = SHA512()
        sha.update(data: raw)
        guard cipherCommitment == Data(sha.finalize()) else { throw NearbyError.ukey2 }

        let clientFinish = try Securegcm_Ukey2ClientFinished(serializedData: msg.messageData)
        guard clientFinish.hasPublicKey else { throw NearbyError.requiredFieldMissing("ukey2clientFinish.publicKey") }
        let clientKey = try Securemessage_GenericPublicKey(serializedData: clientFinish.publicKey)

        try finalizeKeyExchange(peerKey: clientKey)
        currentState = .receivedUkeyClientFinish
    }

    private func processConnectionResponseFrame(_ frame: Location_Nearby_Connections_OfflineFrame) throws {
        guard frame.hasV1, frame.v1.hasType else { throw NearbyError.requiredFieldMissing("offlineFrame.v1.type") }
        if case .connectionResponse = frame.v1.type {
            var resp = Location_Nearby_Connections_OfflineFrame()
            resp.version = .v1
            resp.v1 = Location_Nearby_Connections_V1Frame()
            resp.v1.type = .connectionResponse
            resp.v1.connectionResponse = Location_Nearby_Connections_ConnectionResponseFrame()
            resp.v1.connectionResponse.response = .accept
            resp.v1.connectionResponse.status = 0
            resp.v1.connectionResponse.osInfo = Location_Nearby_Connections_OsInfo()
            resp.v1.connectionResponse.osInfo.type = .apple
            sendFrameAsync(try resp.serializedData())

            encryptionDone = true

            var pairedEncryption = Sharing_Nearby_Frame()
            pairedEncryption.version = .v1
            pairedEncryption.v1 = Sharing_Nearby_V1Frame()
            pairedEncryption.v1.type = .pairedKeyEncryption
            pairedEncryption.v1.pairedKeyEncryption = Sharing_Nearby_PairedKeyEncryptionFrame()
            pairedEncryption.v1.pairedKeyEncryption.secretIDHash = Data.randomData(length: 6)
            pairedEncryption.v1.pairedKeyEncryption.signedData = Data.randomData(length: 72)
            try sendTransferSetupFrame(pairedEncryption)
            currentState = .sentConnectionResponse
        } else {
            print("QS: offline frame plano no manejado: \(frame)")
        }
    }

    private func processPairedKeyEncryptionFrame(_ frame: Sharing_Nearby_Frame) throws {
        guard frame.hasV1, frame.v1.hasPairedKeyEncryption else { throw NearbyError.requiredFieldMissing("v1.pairedKeyEncryption") }
        var pairedResult = Sharing_Nearby_Frame()
        pairedResult.version = .v1
        pairedResult.v1 = Sharing_Nearby_V1Frame()
        pairedResult.v1.type = .pairedKeyResult
        pairedResult.v1.pairedKeyResult = Sharing_Nearby_PairedKeyResultFrame()
        pairedResult.v1.pairedKeyResult.status = .unable
        try sendTransferSetupFrame(pairedResult)
        currentState = .sentPairedKeyResult
    }

    private func processPairedKeyResultFrame(_ frame: Sharing_Nearby_Frame) throws {
        guard frame.hasV1, frame.v1.hasPairedKeyResult else { throw NearbyError.requiredFieldMissing("v1.pairedKeyResult") }
        currentState = .receivedPairedKeyResult
    }

    private func makeFileDestinationURL(_ initialDest: URL) -> URL {
        var dest = initialDest
        if FileManager.default.fileExists(atPath: dest.path) {
            var counter = 1
            var path: String
            let ext = dest.pathExtension
            let baseUrl = dest.deletingPathExtension()
            repeat {
                path = "\(baseUrl.path) (\(counter))"
                if !ext.isEmpty { path += ".\(ext)" }
                counter += 1
            } while FileManager.default.fileExists(atPath: path)
            dest = URL(fileURLWithPath: path)
        }
        return dest
    }

    private func processIntroductionFrame(_ frame: Sharing_Nearby_Frame) throws {
        guard frame.hasV1, frame.v1.hasIntroduction else { throw NearbyError.requiredFieldMissing("v1.introduction") }
        currentState = .waitingForUserConsent
        guard frame.v1.introduction.fileMetadata.count > 0, frame.v1.introduction.textMetadata.isEmpty else {
            rejectTransfer(with: .unsupportedAttachmentType)
            return
        }
        try FileManager.default.createDirectory(at: saveDirectory, withIntermediateDirectories: true)
        for file in frame.v1.introduction.fileMetadata {
            let dest = makeFileDestinationURL(saveDirectory.appendingPathComponent(file.name))
            let info = InternalFileInfo(
                meta: FileMetadata(name: file.name, size: file.size, mimeType: file.mimeType),
                payloadID: file.payloadID,
                destinationURL: dest)
            transferredFiles[file.payloadID] = info
        }
        totalBytes = transferredFiles.values.reduce(0) { $0 + $1.meta.size }
        receivedBytes = 0
        let metadata = TransferMetadata(files: transferredFiles.map { $0.value.meta }, id: id, pinCode: pinCode)
        let device = remoteDeviceInfo!
        DispatchQueue.main.async {
            self.delegate?.obtainUserConsent(for: metadata, from: device, connection: self)
        }
    }

    func submitUserConsent(accepted: Bool) {
        DispatchQueue.global(qos: .utility).async {
            if accepted { self.acceptTransfer() } else { self.rejectTransfer() }
        }
    }

    private func acceptTransfer() {
        do {
            for (id, file) in transferredFiles {
                FileManager.default.createFile(atPath: file.destinationURL.path, contents: nil)
                let handle = try FileHandle(forWritingTo: file.destinationURL)
                transferredFiles[id]!.fileHandle = handle
                transferredFiles[id]!.created = true
            }
            var frame = Sharing_Nearby_Frame()
            frame.version = .v1
            frame.v1.type = .response
            frame.v1.connectionResponse.status = .accept
            currentState = .receivingFiles
            try sendTransferSetupFrame(frame)
        } catch {
            lastError = error
            protocolError()
        }
    }

    private func rejectTransfer(with reason: Sharing_Nearby_ConnectionResponseFrame.Status = .reject) {
        var frame = Sharing_Nearby_Frame()
        frame.version = .v1
        frame.v1.type = .response
        frame.v1.connectionResponse.status = reason
        do {
            try sendTransferSetupFrame(frame)
            try sendDisconnectionAndDisconnect()
        } catch {
            print("QS: error rechazando: \(error)")
            protocolError()
        }
    }

    private func deletePartiallyReceivedFiles() throws {
        for (_, file) in transferredFiles where file.created {
            try? FileManager.default.removeItem(at: file.destinationURL)
        }
    }
}

File: android/app/src/main/java/com/wiwy/wiwytransfer/qs/InboundNearbyConnection.kt
package com.wiwy.wiwytransfer.qs

import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.net.Socket

/** Metadatos de un archivo entrante por Quick Share. */
data class QsFileMeta(val name: String, val size: Long, val mimeType: String)

/** Eventos del receptor (se invocan en hilo de IO; el consumidor salta a main para UI). */
interface InboundDelegate {
    fun onConsent(connection: InboundNearbyConnection, sender: String, pin: String?, files: List<QsFileMeta>)
    fun onProgress(received: Long, total: Long, currentFile: String)
    fun onFinished(savedPaths: List<String>, sender: String)
    fun onClosed(error: String?)
}

/** Receptor Quick Share (rol servidor del UKEY2). Port de InboundNearbyConnection.swift. */
class InboundNearbyConnection(
    socket: Socket,
    private val saveDir: File,
    private val delegate: InboundDelegate,
) : NearbyConnection(socket) {

    private enum class State {
        INITIAL, RECEIVED_CONNECTION_REQUEST, SENT_UKEY_SERVER_INIT, RECEIVED_UKEY_CLIENT_FINISH,
        SENT_CONNECTION_RESPONSE, SENT_PAIRED_KEY_RESULT, RECEIVED_PAIRED_KEY_RESULT,
        WAITING_FOR_CONSENT, RECEIVING_FILES, DISCONNECTED
    }

    private class FileState(
        val meta: QsFileMeta, val dest: File,
        var out: FileOutputStream? = null, var bytes: Long = 0, var created: Boolean = false,
    )

    private var state = State.INITIAL
    private var cipherCommitment: ByteArray? = null
    private var senderName = "Quick Share"
    private val files = LinkedHashMap<Long, FileState>()
    private val savedPaths = ArrayList<String>()
    private var totalBytes = 0L
    private var receivedBytes = 0L

    override fun isServer() = true
    override fun connectionReady() { /* el cliente inicia */ }

    override fun onClosed() {
        // borra archivos a medio recibir
        for (f in files.values) if (f.created) runCatching { f.out?.close(); if (state != State.DISCONNECTED) f.dest.delete() }
        delegate.onClosed(null)
    }

    override fun processReceivedFrame(frameData: ByteArray) {
        try {
            when (state) {
                State.INITIAL -> processConnectionRequest(OfflineFrame.parseFrom(frameData))
                State.RECEIVED_CONNECTION_REQUEST -> {
                    ukeyClientInitMsgData = frameData
                    processUkey2ClientInit(Ukey2Message.parseFrom(frameData))
                }
                State.SENT_UKEY_SERVER_INIT -> processUkey2ClientFinish(Ukey2Message.parseFrom(frameData), frameData)
                State.RECEIVED_UKEY_CLIENT_FINISH -> processConnectionResponse(OfflineFrame.parseFrom(frameData))
                else -> decryptAndProcessReceivedSecureMessage(SecureMessage.parseFrom(frameData))
            }
        } catch (e: Exception) {
            Log.w("WiwyQS", "Error en estado $state: ${e.message}")
            close()
        }
    }

    override fun processTransferSetupFrame(frame: SharingFrame) {
        if (frame.hasV1() && frame.v1.type == SharingFrameType.CANCEL) { sendDisconnectionAndDisconnect(); return }
        when (state) {
            State.SENT_CONNECTION_RESPONSE -> processPairedKeyEncryption()
            State.SENT_PAIRED_KEY_RESULT -> state = State.RECEIVED_PAIRED_KEY_RESULT
            State.RECEIVED_PAIRED_KEY_RESULT -> processIntroduction(frame)
            else -> {}
        }
    }

    private fun processConnectionRequest(frame: OfflineFrame) {
        val info = frame.v1.connectionRequest.endpointInfo.toByteArray()
        if (info.size > 17) {
            val nameLen = info[17].toInt() and 0xFF
            if (info.size >= 18 + nameLen) {
                senderName = String(info, 18, nameLen, Charsets.UTF_8)
            }
        }
        Log.i("WiwyQS", "ConnectionRequest de $senderName")
        state = State.RECEIVED_CONNECTION_REQUEST
    }

    private fun processUkey2ClientInit(msg: Ukey2Message) {
        if (msg.messageType != Ukey2Type.CLIENT_INIT) { sendUkey2Alert(Ukey2AlertType.BAD_MESSAGE_TYPE); return }
        val clientInit = Ukey2ClientInit.parseFrom(msg.messageData)
        var found = false
        for (c in clientInit.cipherCommitmentsList) {
            if (c.handshakeCipher == Ukey2HandshakeCipher.P256_SHA512) { cipherCommitment = c.commitment.toByteArray(); found = true; break }
        }
        if (!found) { sendUkey2Alert(Ukey2AlertType.BAD_HANDSHAKE_CIPHER); return }

        keyPair = Crypto.generateKeyPair()
        val (x, y) = Crypto.publicKeyXY(keyPair!!.public)
        val pkey = GenericPublicKey.newBuilder()
            .setType(PublicKeyTypeEnum.EC_P256)
            .setEcP256PublicKey(EcP256PublicKey.newBuilder().setX(x.toByteString()).setY(y.toByteString()))
            .build()
        val serverInit = Ukey2ServerInit.newBuilder()
            .setVersion(1)
            .setRandom(Crypto.randomBytes(32).toByteString())
            .setHandshakeCipher(Ukey2HandshakeCipher.P256_SHA512)
            .setPublicKey(pkey.toByteString())
            .build()
        val serverInitMsg = Ukey2Message.newBuilder()
            .setMessageType(Ukey2Type.SERVER_INIT)
            .setMessageData(serverInit.toByteString())
            .build()
        ukeyServerInitMsgData = serverInitMsg.toByteArray()
        sendFrame(ukeyServerInitMsgData!!)
        state = State.SENT_UKEY_SERVER_INIT
    }

    private fun processUkey2ClientFinish(msg: Ukey2Message, raw: ByteArray) {
        if (msg.messageType != Ukey2Type.CLIENT_FINISH) { close(); return }
        if (!Crypto.sha512(raw).contentEquals(cipherCommitment)) throw NearbyException("commitment no coincide")
        val finished = Ukey2ClientFinished.parseFrom(msg.messageData)
        val clientKey = GenericPublicKey.parseFrom(finished.publicKey)
        finalizeKeyExchange(clientKey)
        Log.i("WiwyQS", "UKEY2 completado (PIN $pinCode)")
        state = State.RECEIVED_UKEY_CLIENT_FINISH
    }

    private fun processConnectionResponse(frame: OfflineFrame) {
        if (frame.v1.type != ConnFrameType.CONNECTION_RESPONSE) return
        val resp = OfflineFrame.newBuilder()
            .setVersion(OfflineVersion.V1)
            .setV1(
                ConnV1Frame.newBuilder()
                    .setType(ConnFrameType.CONNECTION_RESPONSE)
                    .setConnectionResponse(
                        ConnectionResponseFrame.newBuilder()
                            .setResponse(ConnResponseStatus.ACCEPT)
                            .setStatus(0)
                            .setOsInfo(OsInfo.newBuilder().setType(OsType.ANDROID))
                    )
            )
            .build()
        sendFrame(resp.toByteArray())
        encryptionDone = true

        val pke = SharingFrame.newBuilder()
            .setVersion(SharingVersion.V1)
            .setV1(
                SharingV1Frame.newBuilder()
                    .setType(SharingFrameType.PAIRED_KEY_ENCRYPTION)
                    .setPairedKeyEncryption(
                        PairedKeyEncryptionFrame.newBuilder()
                            .setSecretIdHash(Crypto.randomBytes(6).toByteString())
                            .setSignedData(Crypto.randomBytes(72).toByteString())
                    )
            )
            .build()
        sendTransferSetupFrame(pke)
        state = State.SENT_CONNECTION_RESPONSE
    }

    private fun processPairedKeyEncryption() {
        val pkr = SharingFrame.newBuilder()
            .setVersion(SharingVersion.V1)
            .setV1(
                SharingV1Frame.newBuilder()
                    .setType(SharingFrameType.PAIRED_KEY_RESULT)
                    .setPairedKeyResult(PairedKeyResultFrame.newBuilder().setStatus(PairedKeyResultStatus.UNABLE))
            )
            .build()
        sendTransferSetupFrame(pkr)
        state = State.SENT_PAIRED_KEY_RESULT
    }

    private fun processIntroduction(frame: SharingFrame) {
        val intro = frame.v1.introduction
        if (intro.fileMetadataCount == 0) { rejectTransfer(SharingResponseStatus.UNSUPPORTED_ATTACHMENT_TYPE); return }
        saveDir.mkdirs()
        for (fm in intro.fileMetadataList) {
            val name = sanitize(fm.name)
            val dest = uniqueFile(saveDir, name)
            files[fm.payloadId] = FileState(QsFileMeta(name, fm.size, fm.mimeType), dest)
            totalBytes += fm.size
        }
        state = State.WAITING_FOR_CONSENT
        delegate.onConsent(this, senderName, pinCode, files.values.map { it.meta })
    }

    /** El usuario aceptó/rechazó (llamar desde la UI). */
    fun submitConsent(accept: Boolean) {
        if (!accept) { rejectTransfer(SharingResponseStatus.REJECT); return }
        try {
            for (f in files.values) {
                f.out = FileOutputStream(f.dest)
                f.created = true
            }
            val resp = SharingFrame.newBuilder()
                .setVersion(SharingVersion.V1)
                .setV1(
                    SharingV1Frame.newBuilder()
                        .setType(SharingFrameType.RESPONSE)
                        .setConnectionResponse(SharingConnectionResponseFrame.newBuilder().setStatus(SharingResponseStatus.ACCEPT))
                )
                .build()
            state = State.RECEIVING_FILES
            sendTransferSetupFrame(resp)
        } catch (e: Exception) {
            Log.w("WiwyQS", "Error aceptando: ${e.message}"); close()
        }
    }

    override fun processFileChunk(frame: PayloadTransferFrame) {
        val id = frame.payloadHeader.id
        val f = files[id] ?: return
        val chunk = frame.payloadChunk
        if (!chunk.body.isEmpty) {
            val b = chunk.body.toByteArray()
            f.out?.write(b)
            f.bytes += b.size
            receivedBytes += b.size
            delegate.onProgress(receivedBytes, totalBytes, f.meta.name)
        } else if ((chunk.flags.toInt() and 1) == 1) {
            runCatching { f.out?.close() }
            f.out = null
            savedPaths.add(f.dest.absolutePath)
            files.remove(id)
            if (files.isEmpty()) {
                state = State.DISCONNECTED
                delegate.onFinished(savedPaths.toList(), senderName)
                sendDisconnectionAndDisconnect()
            }
        }
    }

    private fun rejectTransfer(status: SharingResponseStatus) {
        val resp = SharingFrame.newBuilder()
            .setVersion(SharingVersion.V1)
            .setV1(
                SharingV1Frame.newBuilder()
                    .setType(SharingFrameType.RESPONSE)
                    .setConnectionResponse(SharingConnectionResponseFrame.newBuilder().setStatus(status))
            )
            .build()
        runCatching { sendTransferSetupFrame(resp); sendDisconnectionAndDisconnect() }
    }

    private fun sanitize(raw: String): String {
        val base = raw.substringAfterLast('/').substringAfterLast('\\').trim()
        return if (base.isEmpty() || base == "." || base == "..") "archivo" else base
    }

    private fun uniqueFile(dir: File, name: String): File {
        var dest = File(dir, name)
        if (!dest.exists()) return dest
        val dot = name.lastIndexOf('.')
        val base = if (dot > 0) name.substring(0, dot) else name
        val ext = if (dot > 0) name.substring(dot) else ""
        var i = 1
        while (dest.exists()) { dest = File(dir, "$base ($i)$ext"); i++ }
        return dest
    }
}
